gusto/initialisation_tools.py

```diff
@@ -293,117 +305,170 @@ def moist_hydrostatic_balance(state, theta_e, water_t, pi_boundary=Constant(1.0)
         state.logger.warning("default quadrature degree most likely not sufficient for this degree element")
-    quadrature_degree = (5, 5)
-
-    params = {'ksp_type': 'preonly',
-              'ksp_monitor_true_residual': True,
-              'ksp_converged_reason': True,
-              'snes_converged_reason': True,
-              'ksp_max_it': 100,
-              'mat_type': 'aij',
-              'pc_type': 'lu',
-              'pc_factor_mat_solver_type': 'mumps'}
 
     theta0.interpolate(theta_e)
     water_v0.interpolate(water_t)
-    Pi = Function(Vr)
-    epsilon = 0.9  # relaxation constant
-
-    # set up mixed space
-    Z = MixedFunctionSpace((Vt, Vt))
-    z = Function(Z)
-
-    gamma, phi = TestFunctions(Z)
-
-    theta_v, w_v = z.split()
-
-    # give first guesses for trial functions
-    theta_v.assign(theta0)
-    w_v.assign(water_v0)
-
-    theta_v, w_v = split(z)
-
-    # define variables
-    T = thermodynamics.T(state.parameters, theta_v, Pi, r_v=w_v)
-    p = thermodynamics.p(state.parameters, Pi)
-    w_sat = thermodynamics.r_sat(state.parameters, T, p)
-
-    dxp = dx(degree=(quadrature_degree))
-
-    # set up weak form of theta_e and w_sat equations
-    F = (-gamma * theta_e * dxp
-         + gamma * thermodynamics.theta_e(state.parameters, T, p, w_v, water_t) * dxp
-         - phi * w_v * dxp
-         + phi * w_sat * dxp)
-
-    problem = NonlinearVariationalProblem(F, z)
-    solver = NonlinearVariationalSolver(problem, solver_parameters=params)
-
-    theta_v, w_v = z.split()
 
-    Pi_h = Function(Vr).interpolate((p / p_0) ** (R_d / cp))
+    rho_h = Function(Vr)
+    rho_averaged = Function(Vt)
+    rho_broken = Function(FunctionSpace(state.mesh, BrokenElement(Vt.ufl_element())))
+    rho_recoverer = Recoverer(rho_broken, rho_averaged)
+    w_h = Function(Vt)
+    theta_h = Function(Vt)
+    theta_e_test = Function(Vt)
+    delta = 0.8
+
+    # expressions for finding theta0 and water_v0 from theta_e and water_t
+    pie = thermodynamics.pi(state.parameters, rho_averaged, theta0)
+    p = thermodynamics.p(state.parameters, pie)
+    T = thermodynamics.T(state.parameters, theta0, pie, water_v0)
+    r_v_expr = thermodynamics.r_sat(state.parameters, T, p)
+    theta_e_expr = thermodynamics.theta_e(state.parameters, T, p, water_v0, water_t)
+
+    for i in range(40):
+        # solve for rho with theta_vd and w_v guesses
+        compressible_hydrostatic_balance(state, theta0, rho_h, top=top,
+                                         pi_boundary=pi_boundary, water_t=water_t,
+                                         solve_for_rho=True)
+
+        # damp solution
+        rho0.assign(rho0 * (1 - delta) + delta * rho_h)
+
+        theta_e_test.assign(theta_e_expr)
+        if errornorm(theta_e_test, theta_e) < 1e-8:
+            break
+
+        # calculate averaged rho
+        rho_broken.interpolate(rho0)
+        rho_recoverer.project()
+
+        # now solve for r_v
+        for j in range(5):
```

Comment:
Same problem as mentioned above... I don't know where this 5 is coming from (or the 3 in line 357). I'm not quite sure if I like the idea of having three nested while loops. I need to think about this some more. Maybe others will have some comments.

Reply:
I still think you should probably also parametrize the inner loop with some argument (set to a default value in the function args). Doing multiple linear solves is expensive. It would be ideal to be able to tweak this from the top level.

Reply:
Yes, I agree with @thomasgibson  that the user should be able to set the number of iterations in the inner loop too - fine to have a default value though.

Reply:
Hi both, thanks, I've done this now!